pkg/codegen/schema/schema.go

```diff
@@ -925,63 +966,159 @@ func (t *types) bindPrimitiveType(name string) (Type, error) {
+	Version *semver.Version // The version component of the schema ref
+}
 
-			if typ, ok := t.objects[token]; ok {
-				return typ, nil
-			}
-			if typ, ok := t.enums[token]; ok {
-				return typ, nil
-			}
-			typ, ok := t.tokens[token]
+// typeSpecRef contains the parsed fields from a type spec reference.
+type typeSpecRef struct {
+	*externalSchemaRef
+
+	URL   *url.URL // The parsed URL
+	Token string   // The type token
+	Kind  string   // The kind of reference, either 'resources' or 'types'
+}
+
+// Regexs used to parse type spec refs. These are declared at the package scope to avoid repeatedly recompiling them.
+var refPathRegex = regexp.MustCompile(`^/?(?P<package>\w+)/(?P<version>v?\d+\.\d+\.\d+)/schema\.json$`)
+var refFragmentRegex = regexp.MustCompile(`^/(?P<kind>resources|types)/(?P<token>.*:.*:.*)$`)
+
+func parseTypeSpecRef(ref string) (typeSpecRef, error) {
+	parsedURL, err := url.Parse(ref)
+	if err != nil {
+		return typeSpecRef{}, errors.Wrapf(err, "failed to parse ref URL: %s", ref)
+	}
+
+	fragment, err := url.PathUnescape(parsedURL.Fragment)
+	if err != nil {
+		return typeSpecRef{}, errors.Wrapf(err, "failed to unescape fragment: %s", parsedURL.Fragment)
+	}
+
+	fragmentMatch := refFragmentRegex.FindStringSubmatch(fragment)
+	if len(fragmentMatch) != 3 {
+		return typeSpecRef{}, fmt.Errorf("failed to parse fragment: %s", fragment)
+	}
+
+	result := typeSpecRef{
+		URL:   parsedURL,
+		Kind:  fragmentMatch[1],
+		Token: fragmentMatch[2],
+	}
+
+	if len(parsedURL.Path) > 0 {
+		path, err := url.PathUnescape(parsedURL.Path)
+		if err != nil {
+			return typeSpecRef{}, errors.Wrapf(err, "failed to unescape path: %s", parsedURL.Path)
+		}
+
+		pathMatch := refPathRegex.FindStringSubmatch(path)
+		if len(pathMatch) != 3 {
+			return typeSpecRef{}, fmt.Errorf("failed to parse path: %s", path)
+		}
+
+		pkg, versionToken := pathMatch[1], pathMatch[2]
+		version, err := semver.ParseTolerant(versionToken)
+		if err != nil {
+			return typeSpecRef{}, errors.Wrapf(err, "failed to parse package version: %s", versionToken)
+		}
+
+		result.externalSchemaRef = &externalSchemaRef{
+			Package: pkg,
+			Version: &version,
+		}
+	}
+
+	return result, nil
+}
+
+func (t *types) bindTypeSpecRef(spec TypeSpec) (Type, error) {
+	// Explicitly handle built-in types so that we don't have to handle this type of path during ref parsing.
+	switch spec.Ref {
+	case "pulumi.json#/Archive":
+		return ArchiveType, nil
+	case "pulumi.json#/Asset":
+		return AssetType, nil
+	case "pulumi.json#/Json":
+		return JSONType, nil
+	case "pulumi.json#/Any":
+		return AnyType, nil
+	}
+
+	ref, err := parseTypeSpecRef(spec.Ref)
+	if err != nil {
+		return nil, err
+	}
+
+	if ref.externalSchemaRef != nil {
+		pkg, err := t.loader.LoadPackage(ref.externalSchemaRef.Package, ref.externalSchemaRef.Version)
+		if err != nil {
+			return nil, errors.Wrapf(err, "resolving package %v", ref.URL)
+		}
+
+		switch ref.Kind {
+		case "types":
+			typ, ok := t.named[ref.Token]
```

Comment:
Similarly, we don't need to refer to the `named` map here--just the referenced package.